### PR TITLE
refactor: isolate the cache-writing logic from the batch function 

### PR DESCRIPTION
**Pull request by cre8ivejp**
I've refactored the code to isolate the cache-writing logic from the `batch` function, preventing it from executing in multiple workers. Additionally, I've relocated the cache from a local variable to a struct member to ensure each worker operates using the same cache.

----

This pull request introduces changes to the `pkg/eventpersister/persister/persister.go` file, primarily focused on improving the handling of the feature flag last-used cache. The changes include the introduction of a Mutex for synchronization, moving the cache from a local variable to a struct member, and restructuring the cache update and write operations.

The most important changes are:

Cache Management:

* [`import (`](diffhunk://#diff-ad00cb580cd315d8619de51e511dc06052ace57d6146c30b60366e3a97b56d6cR21): Imported the `sync` package to use a Mutex for synchronizing access to the cache.
* [`type Persister struct {`](diffhunk://#diff-ad00cb580cd315d8619de51e511dc06052ace57d6146c30b60366e3a97b56d6cR127-R129): Added `envLastUsedCache` and `mutex` as struct members to `Persister`.
* [`func NewPersister(`](diffhunk://#diff-ad00cb580cd315d8619de51e511dc06052ace57d6146c30b60366e3a97b56d6cR161): Initialized `envLastUsedCache` in the `NewPersister` function.
* [`func (p *Persister) Run() error {`](diffhunk://#diff-ad00cb580cd315d8619de51e511dc06052ace57d6146c30b60366e3a97b56d6cR174): Added a new Goroutine to write the cache.

Cache Update and Write Operations:

* [`func (p *Persister) batch() error {`](diffhunk://#diff-ad00cb580cd315d8619de51e511dc06052ace57d6146c30b60366e3a97b56d6cL194-L198): Removed local cache and its write operations, and replaced them with the struct member cache. [[1]](diffhunk://#diff-ad00cb580cd315d8619de51e511dc06052ace57d6146c30b60366e3a97b56d6cL194-L198) [[2]](diffhunk://#diff-ad00cb580cd315d8619de51e511dc06052ace57d6146c30b60366e3a97b56d6cL208-L213) [[3]](diffhunk://#diff-ad00cb580cd315d8619de51e511dc06052ace57d6146c30b60366e3a97b56d6cL240-L250)
* [`func (p *Persister) cacheLastUsedInfoPerEnv(`](diffhunk://#diff-ad00cb580cd315d8619de51e511dc06052ace57d6146c30b60366e3a97b56d6cL385-R394): Modified this function to use the struct member cache and added Mutex locking for thread safety.
* [`func (p *Persister) cacheEnvLastUsedInfo(`](diffhunk://#diff-ad00cb580cd315d8619de51e511dc06052ace57d6146c30b60366e3a97b56d6cL414-R405): Updated this function to use the struct member cache. [[1]](diffhunk://#diff-ad00cb580cd315d8619de51e511dc06052ace57d6146c30b60366e3a97b56d6cL414-R405) [[2]](diffhunk://#diff-ad00cb580cd315d8619de51e511dc06052ace57d6146c30b60366e3a97b56d6cL442-R454)
* [`func (p *Persister) writeFlagLastUsedInfoCache()`](diffhunk://#diff-ad00cb580cd315d8619de51e511dc06052ace57d6146c30b60366e3a97b56d6cL442-R454): Introduced a new function to write the cache and reset it. This function is run in a separate Goroutine.
* [`func (p *Persister) writeEnvLastUsedInfo(`](diffhunk://#diff-ad00cb580cd315d8619de51e511dc06052ace57d6146c30b60366e3a97b56d6cR470-R471): Modified this function to lock the Mutex before writing the cache and to reset the cache after writing.

### CODE REVIEW

File: pkg/eventpersister/persister/persister.go [hunk 1]
```diff
@@ -439,10 +430,28 @@ func (p *Persister) cacheEnvLastUsedInfo(
...
+			return nil
+		case <-timer.C:
+			p.logger.Debug("Write cache timer triggered")
+			p.writeEnvLastUsedInfo(p.envLastUsedCache)
+			timer.Reset(p.opts.writeCacheInterval)
+		}
+	}
 }
 
 func (p *Persister) writeEnvLastUsedInfo(envCache environmentLastUsedInfoCache) {
```
Comment (kentakozuka) [inline pkg/eventpersister/persister/persister.go:451]:
```suggestion
func (p *Persister) writeEnvLastUsedInfo() {
```

File: pkg/eventpersister/persister/persister.go [hunk 2]
```diff
@@ -382,25 +374,24 @@ func (p *Persister) countUser(key, userID string) error {
...
+			zap.Int("cacheSize", len(p.envLastUsedCache[environmentNamespace])),
 			zap.Int("eventSize", len(events)),
 		)
 	}
 }
 
 func (p *Persister) cacheEnvLastUsedInfo(
 	event *eventproto.EvaluationEvent,
-	envCache environmentLastUsedInfoCache,
+	envLastUsedCache environmentLastUsedInfoCache,
```
Comment (kentakozuka) [inline pkg/eventpersister/persister/persister.go:394]:
```suggestion
```

File: pkg/eventpersister/persister/persister.go [hunk 3]
```diff
@@ -382,25 +374,24 @@ func (p *Persister) countUser(key, userID string) error {
 	return nil
 }
 
-func (p *Persister) cacheLastUsedInfoPerEnv(
-	envEvents environmentEventMap,
-	envCache environmentLastUsedInfoCache,
-) {
+func (p *Persister) cacheLastUsedInfoPerEnv(envEvents environmentEventMap) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
```
Comment (kentakozuka) [inline pkg/eventpersister/persister/persister.go:379]: How about moving these codes to `cacheEnvLastUsedInfo()`?

**Review by kentakozuka — COMMENTED** [review]
Thanks! left comments. PTAL.

**Comment by cre8ivejp** [comment]
> Thanks! left comments. PTAL.

@kentakozuka, done.

**Review by kentakozuka — APPROVED** [review]
LGTM 👍

**Pull request merged**